### PR TITLE
Flexible forms

### LINKED ISSUE #19: Add more flexible form extraction functions
The current form extraction function is highly inflexible, especially with weird forms that have no distinguishing properties. I ran into this problem for both the Zotero and PyPI providers.

Possible solutions to have a more flexible API for `get_form`:
 - Add another function that uses CSS selectors
 - Add keyword argument to pass in an arbitrary function to apply to the soup

Example of how this could work:

```python
def custom_get_form(text, func, **kwargs):
    """Pass in func that works on a soup and outputs input fields"""
    soup = BeautifulSoup(text, "html5lib")
    inputs = func(soup)
    data = {
        i.get("name", ""): i.get("value", "") or "" for i in inputs if i.get("name", "")
    }
    return data

custom_get_form(r.text, lambda s: s.find(id="content").find("form").find_all("input"))
```

This method should be flexible enough to allow the entire API of BeautifulSoup to be used by providers without duplicating the data extraction. Thoughts?

### PR DESCRIPTION
**Pull request by adelq**
Add more flexible and more fully documented form extraction functions.

Also has a minor refactor of providers to use the new form extraction methods if it makes things simpler.

Fixes #19

### CODE REVIEW

File: passrotate/forms.py [hunk 1]
```diff
@@ -1,10 +1,41 @@
 from bs4 import BeautifulSoup
 
+
+def get_form_data(inputs):
+    """Returns data dictionary from list of BeautifulSoup input elements.
+
+    :param list inputs: The list of BeautifulSoup input elements.
+    :returns: dictionary with (name, value) pairs from inputs.
+    :rtype: dict(str, str)
```
Comment (SirCmpwn) [inline passrotate/forms.py]: Rather than :rtype how about type annotations?

**Comment by SirCmpwn** [comment]
LGTM :+1:

File: passrotate/forms.py (hunk 1 above)
Comment (adelq) [reply passrotate/forms.py]: Good idea, let me get that in

**Comment by adelq** [comment]
Type annotations were too big of a rabbit hole, I think having `rtype` with better Sphinx integration is better for the time being. Merge?

**Comment by SirCmpwn** [comment]
I'm not particularly interested in Sphinx. This is Python, we use pydoc.

**Comment by adelq** [comment]
Fair enough, if that's the case then I'll add the type hints and make the docstrings more readable since I was aiming for Sphinx compatibility.

**Comment by SirCmpwn** [comment]
Thanks very much!

**Pull request merged**